File: src/micapolos/zexy/Key.java
package micapolos.zexy;

import static micapolos.tata8.Game.*;
import static micapolos.zexy.Boolean.*;
import static micapolos.zexy.Event.*;
import static micapolos.zexy.Span.*;

public enum Key implements Showable {
  LEFT(keys.left),
  RIGHT(keys.right),
  UP(keys.up),
  DOWN(keys.down),
  Z(keys.z),
  X(keys.x);

  public final Event press;
  public final Event release;
  public final Boolean isPressed;

  Key(micapolos.tata8.Key state) {
    this.press = event(state::pressed);
    this.release = event(state::released);
    this.isPressed = bool(state::isPressed);
  }

  Key(Boolean isPressed) {
    this.isPressed = isPressed;
    this.press = isPressed.changeTo(true);
    this.release = isPressed.changeTo(false);
  }

  public Span pressedSpan() {
    return span(isPressed, press, release);
  }

  @Override
  public String toString() {
    return String.format("Key.%s: %s", name(), isPressed.get() ? "pressed" : "released");
  }

  static void main() {
    Key.LEFT.show();
  }
}

File: src/micapolos/zexy/Key.kt
package micapolos.zexy

import micapolos.zexy.model.Key as ModelKey

class Key(internal val model: ModelKey) {
  companion object {
    val left = Key(ModelKey.LEFT)
    val right = Key(ModelKey.RIGHT)
    val up = Key(ModelKey.UP)
    val down = Key(ModelKey.DOWN)
    val z = Key(ModelKey.Z)
    val x = Key(ModelKey.X)
  }
}

val key = Key.Companion

val Key.isPressed get() = Bool(micapolos.zexy.model.Integer.KeyDown(model))

context(_: Animation.Block)
val Key.press get() = isPressed.changeTo(true)

context(_: Animation.Block)
val Key.release get() = isPressed.changeTo(false)
